codecs/rotate/rotate.rs

```diff
@@ -1,90 +1,82 @@
+        match self {
+            Some(t) => t,
+            None => std::process::abort(),
+        }
+    }
 
-  // In the straight-copy case, d1 is x, d2 is y.
-  // x starts at 0 and increases.
-  // y starts at 0 and increases.
-  let mut d1_start: isize = 0;
-  let mut d1_limit: isize = input_width;
-  let mut d1_advance: isize = 1;
-  let mut d1_multiplier: isize = 1;
-  let mut d2_start: isize = 0;
-  let mut d2_limit: isize = input_height;
-  let mut d2_advance: isize = 1;
-  let mut d2_multiplier: isize = input_width;
+    #[cfg(debug_assertions)]
+    fn unwrap_hard(self) -> T {
+        o.unwrap()
+    }
+}
 
-  if rotate == 90 {
-    // d1 is y, d2 is x.
-    // y starts at its max value and decreases.
-    // x starts at 0 and increases.
-    d1_start = input_height - 1;
-    d1_limit = input_height;
-    d1_advance = -1;
-    d1_multiplier = input_width;
-    d2_start = 0;
-    d2_limit = input_width;
-    d2_advance = 1;
-    d2_multiplier = 1;
-  } else if rotate == 180 {
-    // d1 is x, d2 is y.
-    // x starts at its max and decreases.
-    // y starts at its max and decreases.
-    d1_start = input_width - 1;
-    d1_limit = input_width;
-    d1_advance = -1;
-    d1_multiplier = 1;
-    d2_start = input_height - 1;
-    d2_limit = input_height;
-    d2_advance = -1;
-    d2_multiplier = input_width;
-  } else if rotate == 270 {
-    // d1 is y, d2 is x.
-    // y starts at 0 and increases.
-    // x starts at its max and decreases.
-    d1_start = 0;
-    d1_limit = input_height;
-    d1_advance = 1;
-    d1_multiplier = input_width;
-    d2_start = input_width - 1;
-    d2_limit = input_width;
-    d2_advance = -1;
-    d2_multiplier = 1;
-  }
+const TILE_SIZE: usize = 64;
 
-  let num_pixels = (input_width * input_height) as usize;
-  let in_b: &mut [u32];
-  let out_b: &mut [u32];
-  unsafe {
-    in_b = from_raw_parts_mut::<u32>(4 as *mut u32, num_pixels);
-    out_b = from_raw_parts_mut::<u32>((input_width * input_height * 4 + 4) as *mut u32, num_pixels);
-  }
+#[no_mangle]
+fn rotate(width: usize, height: usize, rotate: usize) {
+    let num_pixels = width * height;
+    let in_b: &mut [u32];
+    let out_b: &mut [u32];
+    unsafe {
+        in_b = from_raw_parts_mut::<u32>(8 as *mut u32, num_pixels);
+        out_b = from_raw_parts_mut::<u32>((num_pixels * 4 + 8) as *mut u32, num_pixels);
+    }
 
-  for d2 in 0..d2_limit {
-    for d1 in 0..d1_limit {
-      let in_idx = (d1_start + d1 * d1_advance) * d1_multiplier + (d2_start + d2 * d2_advance) * d2_multiplier;
-      *unwrap_abort(out_b.get_mut(i as usize)) = *unwrap_abort(in_b.get(in_idx as usize));
-      i += 1;
+    match rotate {
+        0 => {
+            for i in 0..num_pixels {
+                *out_b.get_mut(i).unwrap_hard() = *in_b.get(i).unwrap_hard();
+            }
+        }
+        90 => {
+            let new_width = height;
+            let _new_height = width;
+            for y_start in (0..height).step_by(TILE_SIZE) {
```

Comment:
@bmeurer @mathiasbynens The performance benefits are _real_. Is this something that v8 could do automatically?

Reply:
You mean something that Rust could do automatically?

(I'm maybe missing the relevant context here BTW... so please bear with me)

Reply:
Yeah sorry to just drop you in like that. We learned today that loops can/should be tiled when iterating over large chunks of memory to make better use of the processors L1 cache. 

The difference in the benchmark shows that this is indeed a massive performance gain. Isn't that something that could be optimized at the level of JS ans Wasm, so that developers don't have to implement the tiling by hand?

Reply:
What you are looking for is essentially a compiler pass that does LNO (loop nest optimization). This is most effective when there's a certain knowledge of the data structure (and often makes sense as a manual optimization applied by the programmer, who knows the data structure).

I'd assume that the Rust compiler or LLVM would be in a better position here. The WASM compiler would need to do quite a bit of analysis to reconstruct enough information to even do some kind of educated guess whether it's a good idea or not. Also when you put such heavy optimizations into the WASM JIT, you pay a runtime cost.

Maybe talk to the LLVM/Rust folks about it?

Reply:
Thanks for explaining! I wasn’t necessarily implying that V8 _should_ do this, I figured way smarter than me would have already done so if it was a good idea. But it seems it’s just not that easy to make that decision by looking at the byte code. Thank you again :)

Reply:
@bmeurer The full context is here: https://news.ycombinator.com/item?id=19167078 So the question is not just about enabling this optimization starting from Wasm bytecode, but also for JS source code. I suppose the following applies to JS as well though:

> it’s just not that easy to make that decision by looking at the byte code.

...or the JavaScript source code.

Reply:
Yes, it's pretty much orthogonal to the concrete programming language. And the same goes for the JavaScript JIT. Doing this properly requires some knowledge about the concrete data structure used in the application.